lsp/source/lib/util.mts

```diff
@@ -315,7 +315,7 @@ export function forwardMap(sourcemapLines: SourcemapLines, position: Position) {
         // find best line without going beyond
         const [_p, _f, srcLine, srcOffset] = mapping
         if (srcLine <= origLine) {
-          if (srcLine >= bestLine && (srcOffset <= origOffset) && (srcOffset >= bestOffset)) {
+          if (srcLine > bestLine && (srcOffset <= origOffset) || srcLine === bestLine && (srcOffset <= origOffset) && (srcOffset >= bestOffset)) {
```

Comment:
Should `srcOffset <= origOffset` be moved to the previous `if`, given that it needs to hold in all cases?

I admit I don't understand why `srcLine <= origLine` or `srcOffset <= origOffset` are requirements...

Reply:
The normal direction for sourcemapping is reverse, given a position in the generated file it points to a position in the source file.

To do the opposite and find the position in the generated file from the source file ("forward map") we want to find the "best" mapping. The best mapping is the closest mapping with line <= original line (ideally it will be equal to the original line), and column <= (original column).

To find that mapping we check every reverse mapping holding on to the best one so far. If we're in the middle of an identifier and the best one begins a few characters before the original column that is probably fine since we don't map into the middle of identifiers. We linearly advance the found line and offset by the difference.

It's all slightly complicated but hopefully that explains it adequately.

Reply:
Thanks, that helps. Perhaps this could be added as a comment?